### PR TITLE
Add pagination to fetch screens api

### PR DESCRIPTION
**Pull request by jerrylin-jf**
add pagination to fetch screens api, because the query limit parameter maximum is 100.

https://docs.zeplin.dev/reference#pagination

### CODE REVIEW

**Comment by hearsilent** [comment]
LGTM
I'll review this ASAP.

File: app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt [hunk 1]
```diff
@@ -91,12 +101,6 @@ class ProjectActivity : AppCompatActivity(), View.OnClickListener {
         }
     }
 
-    override fun onResume() {
-        super.onResume()
-        if (mScreenAdapter != null) {
-            mScreenAdapter!!.notifyDataSetChanged()
-        }
-    }
```
Comment (hearsilent) [inline app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt:99]: @jerrylin-jf This is used for reload failed image.

File: app/build.gradle [hunk 2]
```diff
@@ -66,4 +66,6 @@ dependencies {
     testImplementation 'junit:junit:4.12'
     androidTestImplementation 'androidx.test.ext:junit:1.1.1'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
+
+    implementation "androidx.paging:paging-runtime:2.1.2"
```
Comment (hearsilent) [inline app/build.gradle]: Please insert into L46

File: app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt [hunk 3]
```diff
@@ -45,24 +50,22 @@ class ProjectActivity : AppCompatActivity(), View.OnClickListener {
...
-        NetworkHelper.getScreens(this, mProjectModel!!.id, object : ScreensCallback() {
+        val dataSourceFactory = ScreenModelDataSourceFactory(this, mProjectModel!!.id, object : ScreensCallback() {
             override fun onSuccess(screens: List<ScreenModel>) {
                 if (isFinishing) {
                     return
                 }
                 runOnUiThread {
                     swipeRefreshLayout.isRefreshing = false
                     recyclerView.visibility = View.VISIBLE
-                    recyclerView.setHasFixedSize(true)
```
Comment (hearsilent) [inline app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt:63]: Why don't need `setHasFixedSize` anymore?

File: app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt [hunk 4]
```diff
@@ -83,6 +86,13 @@ class ProjectActivity : AppCompatActivity(), View.OnClickListener {
                 }
             }
         })
+        val pagedListConfig =
+            PagedList.Config.Builder().setPageSize(100).setEnablePlaceholders(false).build()
```
Comment (hearsilent) [inline app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt]: Please extract page size to `Constant.kt`.

File: app/src/main/java/hearsilent/zeplin/libs/NetworkHelper.kt [hunk 5]
```diff
@@ -25,7 +25,7 @@ object NetworkHelper {
     private const val ZEPLIN_PROJECTS_URL = "${ZEPLIN_BASE_URL}/projects"
     private const val ZEPLIN_PROJECT_URL = "${ZEPLIN_BASE_URL}/projects/%s"
     private const val ZEPLIN_SCREENS_URL =
-        "${ZEPLIN_BASE_URL}/projects/%s/screens?sort=section&limit=999&offset=0"
+        "${ZEPLIN_BASE_URL}/projects/%s/screens?sort=section&limit=100&offset=%s"
```
Comment (hearsilent) [inline app/src/main/java/hearsilent/zeplin/libs/NetworkHelper.kt]: Please extract limit of page size to `Constant.kt`.

Comment (hearsilent) [reply app/src/main/java/hearsilent/zeplin/libs/NetworkHelper.kt]: Please replace `%s` to `%d`

File: app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    private val pid: String,
+    private val screensCallback: ScreensCallback
+) : PageKeyedDataSource<Int, ScreenModel>() {
+    override fun loadInitial(
+        params: LoadInitialParams<Int>,
+        callback: LoadInitialCallback<Int, ScreenModel>
+    ) {
+        NetworkHelper.getScreens(context, pid, 0, object : ScreensCallback() {
+            override fun onSuccess(screens: List<ScreenModel>) {
+                callback.onResult(screens, null, screens.size)
```
Comment (hearsilent) [inline app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt:21]: Please add note for this line.

File: app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            override fun onFail(errorMessage: String?) {
+                screensCallback.onFail(errorMessage)
+            }
+        })
+    }
+
+    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, ScreenModel>) {
+        NetworkHelper.getScreens(context, pid, params.key, object : ScreensCallback() {
+            override fun onSuccess(screens: List<ScreenModel>) {
+                callback.onResult(screens, params.key + screens.size)
```
Comment (hearsilent) [inline app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt:35]: Please add note for this line.

**Review by hearsilent — CHANGES_REQUESTED** [review]
@jerrylin-jf I've some feedback, hope you can resolve it! Thx again!

File: app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt (hunk 3 above)
Comment (jerrylin-jf) [reply app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt:63]: In my case, sometimes the screen picture size is not fixed, but I think set setHasFixedSize is true that is ok.

File: app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt (hunk 1 above)
Comment (jerrylin-jf) [reply app/src/main/java/hearsilent/zeplin/activity/ProjectActivity.kt:99]: ok

File: app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt (hunk 6 above)
Comment (jerrylin-jf) [reply app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt:21]: ok

File: app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt (hunk 7 above)
Comment (jerrylin-jf) [reply app/src/main/java/hearsilent/zeplin/libs/ScreenModelDataSource.kt:35]: ok

**Review by jerrylin-jf — COMMENTED** [review]
I already modified my code according to your suggestion, thank you for your feedback.

**Comment by hearsilent** [comment]
Thx a lot!

**Pull request merged**